unified_planning/io/ma_pddl_writer.py

```diff
@@ -0,0 +1,1066 @@
+                        out.write(
+                            f'\n   {" ".join([self._get_mangled_name(o) for o in objects])} - {self._get_mangled_name(t)}'
+                        )
+
+            # If agents are not defined as "constant" in the domain, they are defined in the problem
+            if len(self.problem.agents) > 0:
+                for agent in self.problem.agents:
+                    if agent not in self.domain_objects_agents.keys():
+                        out.write(f'\n   {agent.name} - {agent.name + "_type"}')
+                    else:
+                        out.write(f"")
+
+            out.write("\n )\n")
+            converter = ConverterToMAPDDLString(
+                self.problem.env, self._get_mangled_name
+            )
+            out.write(" (:init")
+            for f, v in self.problem.initial_values.items():
+                if v.is_true():
+                    if f.is_dot():
+                        fluent = f.args[0].fluent()
+                        args = f.args
+                        # objects = f.args[0].args
+                        if f.agent().name == ag.name and fluent in ag.fluents:
+                            out.write(
+                                f'\n  (a_{self._get_mangled_name(fluent)}{" "}{ag.name}{" " if len(args) > 0 else ""}{" ".join([converter.convert(obj) for obj in f.args[0].args])})'
+                            )
+                        elif (
+                            f.agent().name != ag.name
+                            and fluent in self.domain_fluents_agents
+                        ):
+                            out.write(f"\n  {converter.convert(f)}")
+                        else:
+                            out.write(f"")
+                    else:
+                        out.write(f"\n  {converter.convert(f)}")
+                elif v.is_false():
+                    pass
+                else:
+                    out.write(f"\n  (= {converter.convert(f)} {converter.convert(v)})")
+            if self.problem.kind.has_actions_cost():
+                out.write(f" (= (total-cost) 0)")
+            out.write(")\n")
+            out.write(f" (:goal (and")
+            for p in self.problem.goals:
+                if p.is_dot():
+                    fluent = p.args[0].fluent()
+                    args = p.args
+                    # objects = p.args[0].args
+                    agent = p.agent()
+                    out.write(
+                        f' (a_{self._get_mangled_name(fluent)}{" "}{agent.name}{" " if len(args) > 0 else ""}{" ".join([converter.convert(obj) for obj in p.args[0].args])})'
+                    )
+                else:
+                    out.write(f" {converter.convert(p)}")
+            out.write(f"))")
+            out.write("\n)")
+            out.seek(0)
+            ag_problem = out.read()
+            ag_problems[ag.name] = ag_problem
+            out.truncate(0)
+            out.seek(0)
```

Comment:
WTF?
Why do you take `out` in input at all? You can simply use a [`StringIO`](https://docs.python.org/3/library/io.html#io.StringIO) for every agent and return their `.getvalue()`...

Reply:
BTW, you are not guaranteed to have `seek` and `truncate` for every type of `IO[str]`, only for files and buffers...

Reply:
Thanks for the feedback, I understand that using one StringIO for each agent and returning their .getvalue() is much easier. Initially I wanted the multi-agent _write_domain() to take the output as a parameter and thus look similar to the single-agent one. 
Done!